### ISSUE TITLE
Dynamic Screen Setup: getComponentForRouteName -> getComponentForRoute?

### ISSUE THREAD

**Issue by miracle2k** [react-community/react-navigation#187]
I spent the last couple of hours trying to understand react-navigation, and there is a thing I struggle with.

In my app, for a various reasons, it's hard to define the routes statically. I prefer something like:

```js
this.props.navigate(ActualScreenComponent, propsToPass)
```

Now I understand react-navigation uses the static route tree to handle nesting, that is, through the predefined names it can identify the navigator in the tree that is responsible. 

Something I came up with is this:

```js
function RenderAnyPage(props) {
  const { params } = props.navigation.state;
  const Component = params.comp;
  return <Component {...props} />
}
RenderAnyPage.navigationOptions = {
  title: (props) => {
    if (props) return;
    const { state } = props;
    return state && state.params.comp.navigationOptions.title(state);
  }
  ....
}

const SimpleApp = StackNavigator(routeConfig, {
    GlobalNav: {screen: RenderAnyPage}
});

this.props.navigation.navigate('GlobalNav', { comp: SomeScreenComponent })}
```

Annoying that I have to proxy all the navigationOptions, but this seems to work decently.

Trying to improve on this, I tried to see if I can make the router handle actions with component classes directly. In other words, I wanted to do this:

```js
this.props.navigation.navigate(SomeScreenComponent);
```

So I tried:

```js
const SimpleApp = StackNavigator(routeConfig);

const oldGetState = SimpleApp.router.getStateForAction;
SimpleApp.router.getStateForAction = (action, state) => {
  let result = oldGetState(action, state);
  if (result == state) {
    // presumably, the route name is a component, thus couldn't be resolved.
    // now I just need to add it to the route stack; but essentially, I can
    // only add names to the route stack.
  }
  return result;
}
```

Look at the comment above. Unfortunately, CardStack, when it ultimately wants to find the screen to render, calls getComponentForRouteName(). I still have to somehow resolve the route by name.

If the function was instead called ``getComponentForRoute``, I could have the reducer add a route with a custom `component` key, and overwrite ``getComponentForRoute`` to consider it.

As it is, the best I could do would be to abuse ``routeName`` and set it to a component anyway, but that seems brittle.

Is there a reason (except maybe that it's too late to change) why ``getComponentForRoute`` isn't a better API? It seems it would be more flexible, while a the same time serving all the use cases.

**Comment by miracle2k** [comment]
Actually, I cannot abuse `routeName` to be a screen instance, because there are invariants in the code that enforce the stringness.

**Comment by satya164** [comment]
Can you outline the reasons for which you can't specify routes statically?

**Comment by grabbou** [comment]
> Unfortunately, CardStack, when it ultimately wants to find the screen to render, calls getComponentForRouteName(). I still have to somehow resolve the route by name

I think you can implement your own hashing strategy if you prefer. 

For example, you can:
```
this.props.navigation.navigate(SomeScreenComponent.routeName);
```
and define routeName as a static on the component.

Then, when `generating` or defining a route map, you can:
```
StackNavigator({
   [SomeScreenComponent.routeName]: { screen: SomeScreenComponent },
});
```

**Comment by miracle2k** [comment]
My app has a pretty flexible screen design. Depending on the customer build, the screens in use vary. This includes for example different implementations of "payment methods", each which can provide it's own screen flow (varying in the number of screens, too).

Like @grabbou suggests, I could solve this by collecting and registering all the screens. I don't like that I have to, but I could.

But for for some screens, I also want to be able to choose between two different navigators. I have two stack navigators. One on the very top of the hierarchy, and one inside a tab. In some cases, I want to configure a customer such that the a certain screen flow is run inside the nested stack, in other that it is run inside the global stack.

Maybe I could solve this too by registering the routes accordingly, I'm not sure. But my preferred design and instinct would be to have the ability to say at navigation time "show screen X in navigator Y". 

I guess I would say this: Having spent a decent amount of time reading the code of react-navigation, I can see that it is designed with a lot of flexibility; and it's *almost* possible to arrange the pieces in a different way to accomplish what I want, but not quite. Apart from the insistence on string-based route names, I noticed;

- Adding custom methods to the navigate property could be much easier.
- Re-using the valuable code in createNavigationContainer() could be easier.
- Customizing the router instance of a Navigator could be easier.

I realize this isn't actionable right now. I might try to play with a fork and see if I can propose some specific patches.

**Comment by abellee** [comment]
@miracle2k 
have you figure your problem? I have the same.

**Comment by kendy37** [comment]
ok...so,whats ur solution? @miracle2k

**Comment by AphexIce** [comment]
Ive followed this exactly and has been very helpful understanding some concepts though I am still a little lost on react-navigation (I have also looked and followed your example for that). I dont understand how I can link to another page from the list view row. Ie like a profile view... Ive tried adding a navigation path to the row file but I am getting "undefined is not an object evaluating _this.props.navigation" I am sure this is to do with the way the navigation is or is not properly getting registered, but again unsure how this should be done properly. I have manged to do it from other stack elements though only if the data is on the exact module.
